// a-promise/Dist/Promise.js


"use strict";

var _createClass = (function () { function defineProperties(target, props) { for (var key in props) { var prop = props[key]; prop.configurable = true; if (prop.value) prop.writable = true; } Object.defineProperties(target, props); } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _classCallCheck = function (instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } };

var Promise = (function () {
  function Promise(Callback) {
    this.OnError = [];
    this.OnSuccess = [];
    this.State = Promise.State.PENDING;

    _classCallCheck(this, Promise);

    if (typeof Callback == "function") {
      setTimeout((function () {
        Callback.call(null, this.resolve.bind(this), this.reject.bind(this));
      }).bind(this), 10);
    }
  }

  Promise.State = { PENDING: 0, FAILURE: 2, SUCCESS: 1 };
  Promise.ThrowErrors = false;

  _createClass(Promise, {
    resolve: {
      value: function resolve(Result) {
        if (this.State !== Promise.State.PENDING) {
          return;
        }
        this.State = Promise.State.SUCCESS;
        this.Result = Result;
        this.OnSuccess.forEach(function (Callback) {
          Callback(Result);
        });
      }
    },
    reject: {
      value: function reject(ErrorMessage) {
        if (this.State !== Promise.State.PENDING) {
          return;
        }
        this.State = Promise.State.FAILURE;
        this.Result = ErrorMessage;
        if (this.OnError.length) {
          this.OnError.forEach(function (Callback) {
            Callback(ErrorMessage);
          });
        } else {
          if (Promise.ThrowErrors) {
            throw new Error("Uncaught Promise Rejection", ErrorMessage);
          }
        }
      }
    },
    "catch": {
      value: function _catch(Callback) {
        if (typeof Callback !== "function") {
          throw new Error("Callback for Promise.catch should be a function");
        }
        if (this.State === Promise.State.PENDING) {
          this.OnError.push(Callback);
        } else if (this.State === Promise.State.FAILURE) {
          Callback(this.Result);
        }
        return this;
      }
    },
    then: {
      value: function then(Callback, OnError) {
        if (typeof Callback !== "function" || OnError && typeof OnError !== "function") {
          throw new Error("Callback(s) for Promise.then should be a function");
        }
        if (this.State === Promise.State.PENDING) {
          this.OnSuccess.push(Callback);
        } else if (this.State === Promise.State.SUCCESS) {
          Callback(this.Result);
        }
        if (OnError) {
          this["catch"](OnError);
        }
        return this;
      }
    }
  }, {
    resolve: {
      value: function resolve(Result) {
        return new Promise(function (resolve) {
          resolve(Result);
        });
      }
    },
    reject: {
      value: function reject(Result) {
        return new Promise(function (resolve, reject) {
          reject(Result);
        });
      }
    },
    all: {
      value: function all(Promises) {
        return new Promise(function (resolve, reject) {
          var ValidPromises = [],
              Status = Promise.State.PENDING,
              ProcessedPromises = 0,
              Results = [];
          Promises.forEach(function (PromiseInst) {
            if (PromiseInst instanceof Promise) {
              ValidPromises.push(PromiseInst);
            }
          });
          if (!ValidPromises.length) {
            return resolve(Results);
          }
          ValidPromises.forEach(function (PromiseInst, Index) {
            PromiseInst.then(function (Result) {
              ProcessedPromises++;
              Results[Index] = Result;
              if (ProcessedPromises === ValidPromises.length) {
                resolve(Results);
              }
            }, function (Error) {
              Status = Promise.State.FAILURE;
              reject(Error);
            });
          });
        });
      }
    }
  });

  return Promise;
})();

module.exports = Promise;
